atlas-spec/mpc-engine/src/party.rs

```diff
@@ -23,49 +28,175 @@ pub enum ProtocolPhase {
+    ///
+    /// This generates the party's global MAC key and sets the protocol phase to
+    /// `PreInit`.
+    pub fn new(
+        channels: ChannelConfig,
+        circuit: &Circuit,
+        logging: bool,
+        mut entropy: Randomness,
+    ) -> Self {
         Self {
+            bit_counter: 0,
             id: channels.id,
+            num_parties: channels.parties.len(),
             channels,
-            global_mac_key: None,
+            global_mac_key: generate_mac_key(&mut entropy).unwrap(),
             circuit: circuit.clone(),
             entropy,
             current_phase: ProtocolPhase::PreInit,
+            log_counter: 0,
+            enable_logging: logging,
         }
     }
 
+    /// Broadcast a `value` and receive other parties' broadcasted values in
+    /// turn.
+    ///
+    /// Broadcast works using the broadcast relay utility:
+    /// - In a first step, compute a cryptographic commitment on `value` and
+    /// send that individually to all other parties.
+    /// - Then send the commitment opening to the broadcast relay.
+    /// - Finally, receive the other parties commitment openings from the
+    ///   broadcast relay and use them to open their commitments to their
+    ///   values.
+    ///
+    /// Since every party can only send one opening value to the broadcast
+    /// relay, which is assumed to be trusted in distributing the received
+    /// openings faithfully, it is ensured that every commitment is opened with
+    /// the opening information. Therefore, because the commitments are binding,
+    /// if a malicious party sends differing commitments to different parties at
+    /// least one opening must fail.
+    ///
+    /// In a non-blackbox way we can also see that parties cannot chose their
+    /// broadcast values dependent on other parties' values: Beside being
+    /// hiding, the Random Oracle commitment we use also does not allow any
+    /// homomorphic operations on commitment values, so even given many
+    /// commitments from other parties a malicious party should not be able to
+    /// chose their commitment in a way that the committed value is dependent on
+    /// other parties' committed values.
+    fn broadcast(&mut self, value: &[u8]) -> Result<Vec<(usize, Vec<u8>)>, Error> {
+        // send/receive commitment to/from all parties
+        let domain_separator = format!("Broadcast-{}", self.id);
+        let (commitment, opening) =
+            Commitment::new(value, domain_separator.as_bytes(), &mut self.entropy)?;
+
+        let mut received_commitments = Vec::new();
+        // Expect earlier parties' commitments.
+        for _i in 0..self.id {
+            let commitment_msg = self.channels.listen.recv().unwrap();
+            if let MessagePayload::BroadcastCommitment(received_commitment) = commitment_msg.payload
+            {
+                assert_eq!(commitment_msg.to, self.id);
```

Comment:
Generally you should use debug asserts so that they don't interfere in release builds.

Reply:
Ah, right! I've changed it across the workspace.